# Pneumonia Detection from X-rays (CNN)

In [ ]:
### Objective

Classify chest X-ray images as pneumonia or normal using a convolutional neural network (CNN).

In [2]:

import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

# Data Generators with corrected path syntax
traingen = ImageDataGenerator(rescale=1./255, rotation_range=15, width_shift_range=0.1,
                              height_shift_range=0.1, shear_range=0.1, zoom_range=0.1, horizontal_flip=True)

valgen = ImageDataGenerator(rescale=1./255)

# Use raw string literals for Windows paths
train_dir = r"C:\Users\Hp\Downloads\chest_xray\train"
val_dir = r"C:\Users\Hp\Downloads\chest_xray\val"
test_dir = r"C:\Users\Hp\Downloads\chest_xray\test"

# Flow from directory with correct path syntax
traindata = traingen.flow_from_directory(train_dir, target_size=(150, 150), batch_size=32, class_mode='binary')
valdata = valgen.flow_from_directory(val_dir, target_size=(150, 150), batch_size=32, class_mode='binary')
testdata = valgen.flow_from_directory(test_dir, target_size=(150, 150), batch_size=32, class_mode='binary', shuffle=False)

# Build model
model = Sequential([
    Conv2D(32, (3,3), activation='relu', input_shape=(150,150,3)),
    MaxPooling2D(2,2),
    Conv2D(64, (3,3), activation='relu'),
    MaxPooling2D(2,2),
    Conv2D(128, (3,3), activation='relu'),
    MaxPooling2D(2,2),
    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.5),
    Dense(1, activation='sigmoid')
])
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train
model.fit(traindata, epochs=3, validation_data=valdata)

# Evaluate
test_loss, test_acc = model.evaluate(testdata)
print("Test accuracy:", test_acc)


Found 5216 images belonging to 2 classes.
Found 16 images belonging to 2 classes.
Found 624 images belonging to 2 classes.


c:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\Hp\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\trainers\data_adapters\py_dataset_adapter.py:121: UserWarning: Your `PyDataset` class should call `super().__init__(**kwargs)` in its constructor. `**kwargs` can include `workers`, `use_multiprocessing`, `max_queue_size`. Do not pass these arguments to `fit()`, as they will be ignored.
  self._warn_if_super_not_called()


Epoch 1/3
163/163 ━━━━━━━━━━━━━━━━━━━━ 153s 928ms/step - accuracy: 0.8020 - loss: 0.4319 - val_accuracy: 0.7500 - val_loss: 0.7009
Epoch 2/3
163/163 ━━━━━━━━━━━━━━━━━━━━ 166s 1s/step - accuracy: 0.8855 - loss: 0.2730 - val_accuracy: 0.6250 - val_loss: 1.2146
Epoch 3/3
163/163 ━━━━━━━━━━━━━━━━━━━━ 155s 948ms/step - accuracy: 0.9072 - loss: 0.2222 - val_accuracy: 0.6250 - val_loss: 1.2233
20/20 ━━━━━━━━━━━━━━━━━━━━ 15s 710ms/step - accuracy: 0.7276 - loss: 0.6589
Test accuracy: 0.7275640964508057


# Summary

- The notebook uses data generators and directory flow for organized image sets (train, val, test).

- The CNN is trained on augmented X-rays and validated after several epochs.

- Training achieved high accuracy; however, validation accuracy fluctuated, and test accuracy settled around 73%.



# Final Insights

- The project demonstrates that a CNN can learn to classify pneumonia from X-ray images, achieving test accuracy of approximately 73%.

- This highlights the promise of deep learning for medical imaging, but also shows challenges due to limited validation data or potential overfitting.

- More data and/or model refinement could improve the reliability of such systems for clinical use.

